File: code/Kotlin/libs/annotation_processing/apt/compiler/src/main/java/inject/processor/creator/BaseMethod.java
package inject.processor.creator;

import com.squareup.javapoet.TypeName;

import java.util.ArrayList;
import java.util.List;

import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.type.TypeMirror;

import inject.annotation.creator.Extra;

/**
 * @author yuansui
 * @since 2020/9/9
 */
abstract class BaseMethod {
    protected CreatorProcessor processor;
    protected Element annotatedElement;

    protected TypeName annotatedTypeName;

    protected List<Element> allFields = new ArrayList<>();
    protected List<Element> optFields = new ArrayList<>();
    protected List<Element> requireFields = new ArrayList<>();


    BaseMethod(CreatorProcessor processor, Element annotatedElement) {
        this.processor = processor;
        this.annotatedElement = annotatedElement;

        annotatedTypeName = processor.getTypeName(annotatedElement);

        getAnnotatedFields(annotatedElement);
        allFields.addAll(optFields);
        allFields.addAll(requireFields);
    }

    private void getAnnotatedFields(Element annotatedElement) {
        for (Element e : annotatedElement.getEnclosedElements()) {
            Extra a = e.getAnnotation(Extra.class);
            if (a != null) {
                if (a.value()) {
                    optFields.add(e);
                } else {
                    requireFields.add(e);
                }
            }
        }

        List<? extends TypeMirror> superTypes = processor.getTypeUtils().directSupertypes(annotatedElement.asType());
        TypeMirror superClassType = superTypes.size() > 0 ? superTypes.get(0) : null;
        Element superClass = superClassType == null ? null : processor.getTypeUtils().asElement(superClassType);
        if (superClass != null && superClass.getKind() == ElementKind.CLASS) {
            getAnnotatedFields(superClass);
        }
    }
}

File: code/Kotlin/libs/annotation_processing/kapt/compiler/src/main/java/inject/compiler/method/creator/BaseMethod.kt
package inject.compiler.method.creator

import inject.annotation.creator.Extra
import inject.compiler.ext.asTypeName
import inject.compiler.processor.BaseProcessor
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind

/**
 * @author yuansui
 * @since 2020/9/9
 */
abstract class BaseMethod<PROCESSOR : BaseProcessor>(
    protected val processor: PROCESSOR,
    protected val annotatedElement: Element
) {
    protected val annotatedTypeName = annotatedElement.asTypeName()

    protected val allFields = mutableListOf<Element>()
    protected val optFields = mutableListOf<Element>()
    protected val requireFields = mutableListOf<Element>()

    init {
        getAnnotatedFields(annotatedElement)
        allFields.addAll(optFields)
        allFields.addAll(requireFields)
    }

    private fun getAnnotatedFields(annotatedElement: Element) {
        for (e in annotatedElement.enclosedElements) {
            val a = e.getAnnotation<Extra>(Extra::class.java)
            if (a != null) {
                if (a.value) {
                    optFields.add(e)
                } else {
                    requireFields.add(e)
                }
            }
        }

        val superTypes = processor.getTypeUtils().directSupertypes(annotatedElement.asType())
        val superClassType = if (!superTypes.isEmpty()) superTypes[0] else null
        val superClass = if (superClassType == null) null else processor.getTypeUtils().asElement(superClassType)
        if (superClass != null && superClass.kind == ElementKind.CLASS) {
            getAnnotatedFields(superClass)
        }
    }
}